### PR TITLE
Feat:Initial implementation of file-upload-tests

### LINKED ISSUE #50: [stencil] tests for file upload interceptor
## Description

There is a file-upload interceptor in the `common` package which helps enable the file upload functionality in a stencil app by just registering the interceptor wherever required. The task is to add tests for this interceptor.

## Goals
- [ ] test for the `file-upload` interceptor

## Implementational Details

Create a test file named `file-upload.interceptor.spec.ts` in the tests directory and write the required test cases using `jest`. Make sure to cover all edge cases.

## Product Name 
Stencil

## Organisation Name
SamagraX

## Tech Skills Needed
TypeScript, Jest, SuperTest, Software Testing

## Mentors
@techsavvyash 

## Complexity
Medium

## Category
Software Testing

# Test Case List
Some of the test cases I would want you to add tests for are:

## Saving to local storage 

1. filename is simple: `test.txt`
2. file name has multiple periods: `text.tar.gz`
3. the service should throw for illegal filenames like: `foo/bar.xz` or `../foo.txt` or `foo/../bar.ext`
4. filename without extensions are allowed `filename`
5. It is throwing if the `STORAGE_ENDPOINT` directory does not exist
6. It is throwing if the `destination` directory does not exist.
7. Allowing for files with spaces in the same `foo bar.ext`
8. Allows for empty `destination` parameter and stores in the root of `STORAGE_ENDPOINT` folder.

### Uploads over HTTP (`process.env.STORAGE_USE_SSL` === false)

1. The returned urls have `http` and port number mentioned
2. All the cases from `# Saving to local storage` while treating `STORAGE_ENDPOINT` as the minio endpoint and `destination` as the bucket name.

### Uploads pver HTTPS (`process.env.STORAGE_USE_SSL` === true)

1. Returned urls have `https` always but only contain port numbers in case of direct `IPv4` or `IPv6` addresses in the `STORAGE_ENDPOINT`. (For example returned url should look like: `https://<IP>:<PORT>/destination/filename`)
2. Returned urls have `https` but not the port number in case the `STORAGE_ENDPOINT` is a proper url such as `cdn-api.dev.samagra.io`. (For example returned url should look like: `https://cdn-api.dev.samagra.io/destination/filename`)
3. All the cases from `# Saving to local storage` while treating `STORAGE_ENDPOINT` as the minio endpoint and `destination` as the bucket name.

### Make-Bucket Endpoint
1. Test the `make-bucket` endpoint for `local storage`
2. Test the `make-bucket` endpoint for `minio`

### Downloads
1. make sure illegal `destination` parameters like `../foo/bar` are rejected right of the back.
2. Make sure test case 1 for `filenames` as well.
3. Make sure we are able to retrieve the file we uploaded correctly when passing legal params.

### Ad-Hoc
- Include test cases for util functions if you feel like some particular case can be tested out and i have misse

### PR DESCRIPTION
**Pull request by vsvishalsharma**
fixes #50 

Tests included:

1. **Test for Instantiation**: 
2. **Test for Handling File Upload**:
3. **Test for Handling Uploaded File**:
4. **Test for Error Handling**:

### CODE REVIEW

**Comment by vsvishalsharma** [comment]
@sooraj1002 can you suggest any changes that i need to make

**Comment by vsvishalsharma** [comment]
@techsavvyash 
shall i write more test cases like handling large file size ,empty file,invalid file format

**Comment by Savio629** [comment]
@vsvishalsharma 
Can you exclude the unnecessary file from the commit( package-lock and yarn lock)?

**Comment by vsvishalsharma** [comment]
@Savio629 done

File: packages/common/test/file-upload.interceptor.spec.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        },
+      ],
+    }).compile();
+    file = { /* mock file object */ };
+    interceptor = module.get(FastifyFileInterceptor);
+  });
+
+  it('should be defined', () => {
+    expect(interceptor).toBeDefined();
+  });
```
Comment (sooraj1002) [inline packages/common/test/file-upload.interceptor.spec.ts:29]: can we make this a little more descriptive

**Comment by sooraj1002** [comment]
@vsvishalsharma could you send a video showing the working tests?

**Review by sooraj1002 — CHANGES_REQUESTED** [review]

**Comment by vsvishalsharma** [comment]
image
@sooraj1002 
sorry for the late response i just wanted to double check everything and i have made some changes in the FastifyFileInterceptor to support the test cases better

**Comment by vsvishalsharma** [comment]
@sooraj1002  kindly ignore jest.unit.json fiel as it is there only for testing purpose

File: packages/common/src/interceptors/file-upload.interceptor.ts
Comment (techsavvyash) [inline packages/common/src/interceptors/file-upload.interceptor.ts:1]: Could you please explain what these changes are doing and why were they required? If you found a bug please feel free to report it by opening a issue ticket.

**Comment by techsavvyash** [comment]
> @sooraj1002  kindly ignore jest.unit.json fiel as it is there only for testing purpose 

Please remove any and all files that are to be ignored.

**Comment by techsavvyash** [comment]
> image
> 
> @sooraj1002 
> 
> sorry for the late response i just wanted to double check everything and i have made some changes in the FastifyFileInterceptor to support the test cases better 

Please make sure to manage and report any bugs that you found by opening an issue ticket and then submit code changes relating to that.

**Comment by vsvishalsharma** [comment]
> > image
> > @sooraj1002
> > sorry for the late response i just wanted to double check everything and i have made some changes in the FastifyFileInterceptor to support the test cases better
> 
> Please make sure to manage and report any bugs that you found by opening an issue ticket and then submit code changes relating to that.

image

the issue is that appmodule.ts is not present in the the directory specified` "../../../packages/user-service/src/app.module"` this is the correct location for it

File: packages/common/src/interceptors/file-upload.interceptor.ts
Comment (techsavvyash) [reply packages/common/src/interceptors/file-upload.interceptor.ts:1]: can you provide a TL;DR :)

Comment (vsvishalsharma) [reply packages/common/src/interceptors/file-upload.interceptor.ts:1]: Sorry my bad the changes are in respect for the test cases .The changes **enhance error handling**, add **file format validation**, and improve test case structure. They make the interceptor more robust, secure, and maintainable.

Comment (vsvishalsharma) [reply packages/common/src/interceptors/file-upload.interceptor.ts:1]: If the test cases of file format and validation and should handle error are not required then the interceptor can be reverted back to the original state but the changes only provide a enhancement over the previous version of the ineterceptor 
```
it('should handle an invalid file format', async () => {
    const file = { originalname: 'test.txt', mimetype: 'text/plain' };
    const context = createMockContext(file);
    const nextHandler = createMockNextHandler();

    await expect(interceptor.intercept(context, nextHandler)).rejects.toThrow('Invalid file format');
  });
```

**Comment by vsvishalsharma** [comment]
image
the test are working fine waiting for your response on the interceptor part shall i change it or not

**Comment by vsvishalsharma** [comment]
hey @techsavvyash @sooraj1002  looking forward for your review

File: packages/user-service/yarn.lock
Comment (techsavvyash) [inline packages/user-service/yarn.lock:1]: revert this change

File: packages/common/src/interceptors/file-upload.interceptor.ts
Comment (techsavvyash) [inline packages/common/src/interceptors/file-upload.interceptor.ts:1]: revert these changes and open a issue ticket if you have found a bug in the component, so that we can properly test and architect the fix for it.

File: packages/common/test/file-upload.interceptor.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (techsavvyash) [reply packages/common/test/file-upload.interceptor.spec.ts:15]: don't nitpick

File: packages/common/test/file-upload.interceptor.spec.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    const context = createMockContext(undefined);
+    const nextHandler = createMockNextHandler();
+
+    await interceptor.intercept(context, nextHandler);
+
+    expect(context.switchToHttp().getRequest().file).toBeUndefined();
+    expect(nextHandler.handle).toHaveBeenCalled();
+  });
+
+  it('should handle getting an uploaded file when file is null', async () => {
```
Comment (techsavvyash) [inline packages/common/test/file-upload.interceptor.spec.ts:56]: why are we allowing uploaded files to be null? isn't this the same as getting files when they are not present?

File: packages/common/test/file-upload.interceptor.spec.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    const context = createMockContext(null);
+    const nextHandler = createMockNextHandler();
+
+    await interceptor.intercept(context, nextHandler);
+
+    expect(context.switchToHttp().getRequest().file).toBeNull();
+    expect(nextHandler.handle).toHaveBeenCalled();
+  });
+
+  it('should handle an invalid file format', async () => {
```
Comment (techsavvyash) [inline packages/common/test/file-upload.interceptor.spec.ts:66]: why are we not allowing `text/plain` as a file format? the interceptor as of now has no restrictions on the formats of the files which can be uploaded.

**Review by techsavvyash — CHANGES_REQUESTED** [review]

**Comment by techsavvyash** [comment]
> hey @techsavvyash @sooraj1002 looking forward for your review

have added a few comments please resolve them and we can close this PR out.

**Comment by vsvishalsharma** [comment]
@techsavvyash sir revereted back the interceptor
